File: pie chart.py
#filtro oc_data_inizio_progetto 2014-1024
#filtra prev_studio_fatt sameyear, early, delayed
import pandas as pd
all_project = pd.read_csv("../open_coesione.csv")
date_columns = ['DATA_INIZIO_PREV_STUDIO_FATT','OC_DATA_INIZIO_PROGETTO','OC_DATA_FINE_PROGETTO_EFFETTIVA']
all_project['DATA_INIZIO_PREV_STUDIO_FATT'] = pd.to_datetime(all_project['DATA_INIZIO_PREV_STUDIO_FATT'])
all_project['OC_DATA_INIZIO_PROGETTO'] = pd.to_datetime(all_project['OC_DATA_INIZIO_PROGETTO'])
all_project['OC_DATA_FINE_PROGETTO_EFFETTIVA'] = pd.to_datetime(all_project['OC_DATA_FINE_PROGETTO_EFFETTIVA'])
time_section = all_project.loc[(all_project["OC_DATA_INIZIO_PROGETTO"].dt.year >= 2014) & (all_project["OC_DATA_INIZIO_PROGETTO"].dt.year <= 2024)]
time_section




# Helper: normalize region names 
def normalize_region_name(name: str) -> str:
    if not isinstance(name, str):
        return ""
    normalized = "".join(ch for ch in name.lower() if ch.isalnum())
    if "trentino" in normalized or "bolzano" in normalized or "altoadige" in normalized:
        return "trentinoaltoadige"
    elif "friuli" in normalized:
        return "friuliveneziagiulia"
    elif "valledaosta" in normalized or "aoste" in normalized:
        return "valledaosta"
    elif "emiliaromagna" in normalized:
        return "emiliaromagna"
    elif "lombardia" in normalized:
        return "lombardia"
    elif "piemonte" in normalized:
        return "piemonte"
    elif "veneto" in normalized:
        return "veneto"
    elif "liguria" in normalized:
        return "liguria"
    elif "toscana" in normalized:
        return "toscana"
    elif "umbria" in normalized:
        return "umbria"
    elif "marche" in normalized:
        return "marche"
    elif "lazio" in normalized:
        return "lazio"
    elif "abruzzo" in normalized:
        return "abruzzo"
    elif "molise" in normalized:
        return "molise"
    elif "campania" in normalized:
        return "campania"
    elif "puglia" in normalized:
        return "puglia"
    elif "basilicata" in normalized:
        return "basilicata"
    elif "calabria" in normalized:
        return "calabria"
    elif "sicilia" in normalized:
        return "sicilia"
    elif "sardegna" in normalized:
        return "sardegna"
    return normalized

#  Apply normalization 
time_section["REGION_KEY"] = time_section["DEN_REGIONE"].apply(normalize_region_name)

# Map normalized keys back to Italian names 
region_name_map = {
    "piemonte": "Piemonte",
    "valledaosta": "Valle d’Aosta",
    "lombardia": "Lombardia",
    "trentinoaltoadige": "Trentino-Alto Adige",
    "veneto": "Veneto",
    "friuliveneziagiulia": "Friuli-Venezia Giulia",
    "liguria": "Liguria",
    "emiliaromagna": "Emilia-Romagna",
    "toscana": "Toscana",
    "umbria": "Umbria",
    "marche": "Marche",
    "lazio": "Lazio",
    "abruzzo": "Abruzzo",
    "molise": "Molise",
    "campania": "Campania",
    "puglia": "Puglia",
    "basilicata": "Basilicata",
    "calabria": "Calabria",
    "sicilia": "Sicilia",
    "sardegna": "Sardegna"
}
time_section["DEN_REGIONE_CLEAN"] = time_section["REGION_KEY"].map(region_name_map)
time_section




on_time = time_section.loc[(time_section["OC_DATA_INIZIO_PROGETTO"].dt.year == time_section["DATA_INIZIO_PREV_STUDIO_FATT"].dt.year)]
# on_time 1284 progetti
delayed = time_section.loc[(time_section["OC_DATA_INIZIO_PROGETTO"].dt.year > time_section["DATA_INIZIO_PREV_STUDIO_FATT"].dt.year)]
# delayed 908 progetti
early = time_section.loc[(time_section["OC_DATA_INIZIO_PROGETTO"].dt.year < time_section["DATA_INIZIO_PREV_STUDIO_FATT"].dt.year)]
# early 580 progetti




from matplotlib import pyplot as plt

pie_data = [len(on_time), len(delayed), len(early)]
pie_labels = ["On Time", "Delayed", "Early"]
total = sum(pie_data)

fig, ax = plt.subplots(figsize=(8, 6))
wedges, texts, autotexts = ax.pie(pie_data,
                               labels=pie_labels,
                               autopct='%1.0f%%',
                                wedgeprops=dict(width=0.7, edgecolor='w'),
                                colors = ["#2A63AD", "#322AAD", "#2AA5AD"]
                                 )
ax.legend(wedges,            
    pie_labels,        
    title="Project start",
    loc="center left",
    bbox_to_anchor=(1, 0.3, 0.5, 1)
)
ax.text(0, 0, f"Total: {total}", ha='center', va='center', fontsize=10)
ax.set_title("Projects start timeliness")
plt.show()

File: assets/js/piechart.js
// project_timeliness.js
// Requires: plotly-2.x.x.min.js, papaparse.min.js

// --- Load and process the CSV file ---
async function loadAndProcessData(csvPath) {
    console.log("📂 Loading CSV:", csvPath);

    const response = await fetch(csvPath);
    const csvText = await response.text();

    const allProjects = Papa.parse(csvText, { header: true, skipEmptyLines: true }).data;
    console.log("✅ Loaded rows:", allProjects.length);

    // --- Convert date columns ---
    const dateColumns = [
        "DATA_INIZIO_PREV_STUDIO_FATT",
        "OC_DATA_INIZIO_PROGETTO",
        "OC_DATA_FINE_PROGETTO_EFFETTIVA"
    ];

    allProjects.forEach(row => {
        dateColumns.forEach(col => {
            if (row[col]) {
                const date = new Date(row[col]);
                row[col] = isNaN(date) ? null : date;
            } else {
                row[col] = null;
            }
        });
    });

    // --- Filter between 2014–2024 ---
    const timeSection = allProjects.filter(row => {
        const start = row["OC_DATA_INIZIO_PROGETTO"];
        return start && start.getFullYear() >= 2014 && start.getFullYear() <= 2024;
    });
    console.log("✅ Filtered rows:", timeSection.length);

    // --- Normalize region names ---
    function normalizeRegionName(name) {
        if (typeof name !== "string") return "";
        let normalized = name.toLowerCase().replace(/[^a-z0-9]/g, "");
        if (normalized.includes("trentino") || normalized.includes("bolzano") || normalized.includes("altoadige"))
            return "trentinoaltoadige";
        if (normalized.includes("friuli")) return "friuliveneziagiulia";
        if (normalized.includes("valledaosta") || normalized.includes("aoste")) return "valledaosta";
        if (normalized.includes("emiliaromagna")) return "emiliaromagna";
        if (normalized.includes("lombardia")) return "lombardia";
        if (normalized.includes("piemonte")) return "piemonte";
        if (normalized.includes("veneto")) return "veneto";
        if (normalized.includes("liguria")) return "liguria";
        if (normalized.includes("toscana")) return "toscana";
        if (normalized.includes("umbria")) return "umbria";
        if (normalized.includes("marche")) return "marche";
        if (normalized.includes("lazio")) return "lazio";
        if (normalized.includes("abruzzo")) return "abruzzo";
        if (normalized.includes("molise")) return "molise";
        if (normalized.includes("campania")) return "campania";
        if (normalized.includes("puglia")) return "puglia";
        if (normalized.includes("basilicata")) return "basilicata";
        if (normalized.includes("calabria")) return "calabria";
        if (normalized.includes("sicilia")) return "sicilia";
        if (normalized.includes("sardegna")) return "sardegna";
        return normalized;
    }

    const regionNameMap = {
        "piemonte": "Piemonte",
        "valledaosta": "Valle d’Aosta",
        "lombardia": "Lombardia",
        "trentinoaltoadige": "Trentino-Alto Adige",
        "veneto": "Veneto",
        "friuliveneziagiulia": "Friuli-Venezia Giulia",
        "liguria": "Liguria",
        "emiliaromagna": "Emilia-Romagna",
        "toscana": "Toscana",
        "umbria": "Umbria",
        "marche": "Marche",
        "lazio": "Lazio",
        "abruzzo": "Abruzzo",
        "molise": "Molise",
        "campania": "Campania",
        "puglia": "Puglia",
        "basilicata": "Basilicata",
        "calabria": "Calabria",
        "sicilia": "Sicilia",
        "sardegna": "Sardegna"
    };

    timeSection.forEach(row => {
        const key = normalizeRegionName(row["DEN_REGIONE"]);
        row["REGION_KEY"] = key;
        row["DEN_REGIONE_CLEAN"] = regionNameMap[key] || key;
    });

    // --- Categorize projects by timeliness ---
    const onTime = timeSection.filter(row =>
        row["OC_DATA_INIZIO_PROGETTO"] &&
        row["DATA_INIZIO_PREV_STUDIO_FATT"] &&
        row["OC_DATA_INIZIO_PROGETTO"].getFullYear() === row["DATA_INIZIO_PREV_STUDIO_FATT"].getFullYear()
    );
    const delayed = timeSection.filter(row =>
        row["OC_DATA_INIZIO_PROGETTO"] &&
        row["DATA_INIZIO_PREV_STUDIO_FATT"] &&
        row["OC_DATA_INIZIO_PROGETTO"].getFullYear() > row["DATA_INIZIO_PREV_STUDIO_FATT"].getFullYear()
    );
    const early = timeSection.filter(row =>
        row["OC_DATA_INIZIO_PROGETTO"] &&
        row["DATA_INIZIO_PREV_STUDIO_FATT"] &&
        row["OC_DATA_INIZIO_PROGETTO"].getFullYear() < row["DATA_INIZIO_PREV_STUDIO_FATT"].getFullYear()
    );

    console.log("📊 On Time:", onTime.length);
    console.log("📊 Delayed:", delayed.length);
    console.log("📊 Early:", early.length);

    // --- Pie chart data ---
    const pieData = [onTime.length, delayed.length, early.length];
    const pieLabels = ["On Time", "Delayed", "Early"];
    const total = pieData.reduce((a, b) => a + b, 0);

    const data = [{
        type: "pie",
        labels: pieLabels,
        values: pieData,
        textinfo: "label+percent",
        insidetextorientation: "radial",
        hole: 0.4,
        marker: {
            colors: ["#2A63AD", "#322AAD", "#2AA5AD"]
        }
    }];

    const layout = {
        title: "Fig.10: Projects' Start Timeliness",
        annotations: [
            {
                text: `Total: ${total}`,
                x: 0.5,
                y: 0.5,
                font: { size: 14 },
                showarrow: false
            }
        ],
        legend: {
            x: 1,
            y: 0.5,
            title: { text: "Project start" }
        },
        margin: { l: 20, r: 100, t: 50, b: 20 }
    };

    Plotly.newPlot("piechart", data, layout);
}

// --- Run when page loads ---
document.addEventListener("DOMContentLoaded", () => {
    loadAndProcessData("data/open_coesione.csv");
});
